Improve Javadoc on component support

ZheSun88 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-notification-flow/8)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/vaadin-notification-flow?pullRequest=8) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/vaadin-notification-flow?pullRequest=8) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/vaadin-notification-flow?pullRequest=8) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/vaadin-notification-flow?pullRequest=8) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/vaadin-notification-flow?pullRequest=8) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/vaadin-notification-flow?pullRequest=8) it.</sub>

denis-anisimov commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 55 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2TU8zCqtSwksNSdA:-L1N2TU8zCqtSwksNSdB:bwe86bs) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L55)):*
>  Note: Component support constructor will NOT work for string parameter
>      * and {@link #setContent(String)} method

I don't think I understand what it means.

The statement needs to be improved.
And I think there should be a check for condition you mention and throw at least in runtime if this condition satisfied.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 146 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2Xo6xkakUENz_PBC:-L1N2Xo6xkakUENz_PBD:b-fpdwv2) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L146)):*
> NOTE: this method will NOT work for component support constructors like
>      * {@link #Notification()} and {@link #Notification(Component...)}

Not sure whether I understand this either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 55 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2TU8zCqtSwksNSdA:-L1NWRIO1lld1Pnc73z1:b3dcsd0) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L55)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  Note: Component support constructor will NOT work for string parameter
>      * and {@link #setContent(String)} method

I don't think I understand what it means.

The statement needs to be improved.
And I think there should be a check for condition you mention and throw at least in runtime if this condition satisfied.
</blockquote></details>

aww. i have change the `setContent()` method to `setText()`, as it is quite confused for the users. 

The default constructor and constructor with add components are component supported, the others are not. I am trying to state this clear for the users, it would be great if you can help. 

 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 55 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2TU8zCqtSwksNSdA:-L1Qj8TvQq1lfopCNHZM:besy45j) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L55)):*
<details><summary><i>Previously, ZheSun88 (Sun Zhe) wrote…</i></summary><blockquote>

aww. i have change the `setContent()` method to `setText()`, as it is quite confused for the users. 

The default constructor and constructor with add components are component supported, the others are not. I am trying to state this clear for the users, it would be great if you can help. 

 
</blockquote></details>

The problem is that I really don't understand what you are trying to express here and in the `setText()` method.

`This constructor will NOT take HTML or text as content.`

The CTOR doesn't take any text or content. So what does it means ?

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 146 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2Xo6xkakUENz_PBC:-L1QjYXTJ2Q6YnPnm5HR:b59jks7) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L146)):*
> This method will be overwritten when mixed with component support
>      * constructors


? 
I don't understand. Who will overwrite the method ?
Are you sure that it's the method will be overwritten ?

May be "innerrHTML" property will be overwritten ?
Will it be overwritten actually ?

I don't see why I can't call this method. Apparently I can. 
What will be the result ?

Are you trying to say that it won't take any effect if the CTOR with array of `Component`s is used ?
Then please say it exactly in this way. And use a javadoc link to the CTOR instead of " component support constructors" expression ( which I don't really understand).

Other two things:
- Are you sure that it needs to be done in this way ? `Dialog` also had `Dialog(String)` CTOR and now it doesn't have: https://github.com/vaadin/vaadin-dialog-flow/blob/master/src/main/java/com/vaadin/flow/component/dialog/Dialog.java
- I would like to get `IllegalStateException` if I use mentioned CTOR and this method to be notified that I'm doing something wrong. Always remember one thing: nobody reads javadocs. If something should not be done then it has to be prevented either via API or via runtime exception. You can't prevent this specific issue via API then please throw an exception.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 55 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2TU8zCqtSwksNSdA:-L1QqDnDPIR5JUwcsCAm:bv3vg95) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L55)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

The problem is that I really don't understand what you are trying to express here and in the `setText()` method.

`This constructor will NOT take HTML or text as content.`

The CTOR doesn't take any text or content. So what does it means ?
</blockquote></details>

I took the https://github.com/vaadin/flow/blob/fc04c2d0ce424ae4fe0062588774b16610e4cf38/flow-server/src/main/java/com/vaadin/flow/component/HasText.java 
as an reference. So i hope it can make it clear. this time. 

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 146 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1N2Xo6xkakUENz_PBC:-L1QqbKcj21KEBrE5100:bktdcbf) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/580bee7ada7b8d086d94421e552c00385054b60c/src/main/java/com/vaadin/flow/component/notification/Notification.java#L146)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> This method will be overwritten when mixed with component support
>      * constructors


? 
I don't understand. Who will overwrite the method ?
Are you sure that it's the method will be overwritten ?

May be "innerrHTML" property will be overwritten ?
Will it be overwritten actually ?

I don't see why I can't call this method. Apparently I can. 
What will be the result ?

Are you trying to say that it won't take any effect if the CTOR with array of `Component`s is used ?
Then please say it exactly in this way. And use a javadoc link to the CTOR instead of " component support constructors" expression ( which I don't really understand).

Other two things:
- Are you sure that it needs to be done in this way ? `Dialog` also had `Dialog(String)` CTOR and now it doesn't have: https://github.com/vaadin/vaadin-dialog-flow/blob/master/src/main/java/com/vaadin/flow/component/dialog/Dialog.java
- I would like to get `IllegalStateException` if I use mentioned CTOR and this method to be notified that I'm doing something wrong. Always remember one thing: nobody reads javadocs. If something should not be done then it has to be prevented either via API or via runtime exception. You can't prevent this specific issue via API then please throw an exception.
</blockquote></details>

there is a discussion about this usage. But for dialog, the common use case is using component in it.. But notification is kind different. often is shows only static text.  That is why we still keep the String constructor. 

If the javadoc is okay, i will add a `IllegalStateException` immediately. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 55 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1QtcfrknpQ_hmYbV_g:-L1QtcfrknpQ_hmYbV_h:b-9jid6a) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L55)):*
> the{@link Text}

space between the word and the link

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 56 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Qtri1ASQb-rZbjyLO:-L1Qtri1ASQb-rZbjyLP:bxem3he) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L56)):*
> {@link #setText(String)} will NOT take effect in this constructor.

Still bad statement.
If I understand right then `setText`  won't work for the component created with this CTOR.
Your statement says that something doesn't wok with the CTOR itself. I don't expect it works in CTOR.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 132 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1QuCpTBrlyGsspaCsc:-L1QuCpTBrlyGsspaCsd:b3nhgyx) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L132)):*
> {@link #setText(String)} will NOT take effect in this constructor.

Same here.
CTOR is not a target of the statement.
A component created with this CTOR is the target.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 144 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1QuTC4pUjXmw4Txx-Q:-L1QuTC4pUjXmw4Txx-R:bx9uvpq) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L144)):*
> innerHTML

I don't think it's good to mention `innerHtml` attribute here.
Technically it doesn't matter how the content/text is set. `innerHtml`  is an implementation detail which we may change in the future. And we should not change javadoc and API if we change the implementation only.

I don't know what's better: text or content but `innerHtml`  is definitely bad.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 147 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Quw3_HtEfA6tNQ2Lo:-L1Quw3a114PuenLw8kQ:b-8rplru) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L147)):*
> {@link #Notification()} and
>      * {@link #Notification(Component...)}. 

in (or for ) component created using ...... constructors.
Object (instances) and CTORs are not the same things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
@denis-anisimov for now, i have changed the code for `setText()` and `add(Component)`, while add a few more IT tests for the functional changes. 

Our user can mix them together, according to Pekka said, but we wont support this change, the component will remove everything and add what the last method does.  I have wrote it in the Javadoc, as throwing an error to the user is not wanted.

sorry to take you so long time and thanks for correcting. 

---

Review status: 1 of 6 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-:-L1SOiqenJhDlZS-kD23:b-djwjbc)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Review status: 1 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 55 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1QtcfrknpQ_hmYbV_g:-L1SYCSHlaEmmsqOjeOn:bb7f2mf) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L55)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> the{@link Text}

space between the word and the link
</blockquote></details>

done

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 56 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Qtri1ASQb-rZbjyLO:-L1SYFqAomvUDiC1lLPH:b-2wysj2) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L56)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@link #setText(String)} will NOT take effect in this constructor.

Still bad statement.
If I understand right then `setText`  won't work for the component created with this CTOR.
Your statement says that something doesn't wok with the CTOR itself. I don't expect it works in CTOR.
</blockquote></details>

as the code change, this part has been removed.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 132 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1QuCpTBrlyGsspaCsc:-L1SYKBJPVD07BsEupUT:b-ygp5d0) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L132)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@link #setText(String)} will NOT take effect in this constructor.

Same here.
CTOR is not a target of the statement.
A component created with this CTOR is the target.
</blockquote></details>

as the code change, this part has been removed

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 144 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1QuTC4pUjXmw4Txx-Q:-L1SYNgrixvpTqSPYryk:bvatfho) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L144)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> innerHTML

I don't think it's good to mention `innerHtml` attribute here.
Technically it doesn't matter how the content/text is set. `innerHtml`  is an implementation detail which we may change in the future. And we should not change javadoc and API if we change the implementation only.

I don't know what's better: text or content but `innerHtml`  is definitely bad.
</blockquote></details>

i chose text. and changed

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 147 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Quw3_HtEfA6tNQ2Lo:-L1SYU6boeS0-h92DwVf:bse1gyh) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/main/java/com/vaadin/flow/component/notification/Notification.java#L147)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@link #Notification()} and
>      * {@link #Notification(Component...)}. 

in (or for ) component created using ...... constructors.
Object (instances) and CTORs are not the same things.
</blockquote></details>

sentence refactored

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 4 of 5 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 231 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WIDu_nFQV1aFXoD6x:-L1WIDu_nFQV1aFXoD6y:b-o6c68r) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/main/java/com/vaadin/flow/component/notification/Notification.java#L231)):*
> templateElement.setProperty("innerHTML", null);

I think there should be a method `removeProperty()`. It's better to use it here.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java, line 133 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJxbkmNJz_ZEvM_8a:-L1WJxbkmNJz_ZEvM_8b:b-x4z0oc) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java#L133)):*
> n

one letter variable.
Don't use one or two letter variables.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java, line 166 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WKDd1xwILWOaE3vtI:-L1WKDd1xwILWOaE3vtJ:b48ae6j) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java#L166)):*
> private void createNotificationAddMix() {
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>
>         Notification n = new Notification();
>         n.setId("add-Mix");
>         NativeButton button1 = new NativeButton(BUTTON_CAPTION);
>         NativeButton button2 = new NativeButton("3333");
>         NativeButton button3 = new NativeButton("4444");
>         NativeButton button4 = new NativeButton("BYE");
>         button1.setId("Add-Mix-open");
>         button2.setId("add-Mix-two");
>         button3.setId("add-Mix-three");
>         button4.setId("add-Mix-close");
</details>
>  
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>
>         n.add(button2);
>         n.setText("55555555");
>         n.add(button3);
>         button1.addClickListener(event -> n.open());
>         button4.addClickListener(event -> n.close());
>         add(n, button1, button4);
>     }
</details>


You have tree new methods here but the IT test does check the same thing using different identifiers.

I don't see in the tests checks for text presence in the dialog.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 112 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJV2gD-9I-p4CBlbj:-L1WJV2gD-9I-p4CBlbk:bndx3g6) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L112)):*
> NotificationAddTwoComponents

use camel case method names even for test methods (the first letter).

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 122 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJelCPFOz38SlIJmY:-L1WJelCPFOz38SlIJmZ:b-42bmxk) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L122)):*
>  findElement(By.id("Add-two-components-open")).click();
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>
>         waitUntil(driver -> Boolean.TRUE.toString()
>                 .equals(findElement(By.tagName(DIALOG_OVERLAY_TAG))
>                         .getAttribute("opened")));
>         Assert.assertEquals(2,
>                 getOverlayContent().findElements(By.tagName("button")).size());
>         clickElementWithJs(findElement(By.id("add-two-components-close")));
>         waitUntil(driver -> Boolean.FALSE.toString()
>                 .equals(findElement(By.tagName(DIALOG_OVERLAY_TAG))
>                         .getAttribute("opened")));
</details>


All three new methods have the same source code except two String identifiers which can be used as parameters in the extracted method.

Plus the number of button in the dialog. It also may be a parameter.

Don't duplicate the code. Extract the logic into a method and use parameters there.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 126 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJaaxrjMvYB_CRN4e:-L1WJaaxrjMvYB_CRN4f:b9wo3ul) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L126)):*
> NotificationAddMix

same here

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 140 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJc7qkuWDe0vWKRdZ:-L1WJc7qkuWDe0vWKRd_:b-wric9w) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L140)):*
> NotificationwithTextAndAddComponent

same here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Reviewed 1 of 2 files at r2, 5 of 5 files at r3.
Review status: 1 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 231 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WIDu_nFQV1aFXoD6x:-L1W_0_KRsaEqqphW0Ts:bi1jyaf) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/main/java/com/vaadin/flow/component/notification/Notification.java#L231)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> templateElement.setProperty("innerHTML", null);

I think there should be a method `removeProperty()`. It's better to use it here.
</blockquote></details>

this line has been removed as `innerHTML` will overwrite the content.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java, line 133 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJxbkmNJz_ZEvM_8a:-L1WZswVuhtFRH1SNB-b:b8boskd) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java#L133)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> n

one letter variable.
Don't use one or two letter variables.
</blockquote></details>

oops. sorry, changed

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java, line 166 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WKDd1xwILWOaE3vtI:-L1WZorsZDNtZ8MozJwb:b6835cs) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java#L166)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private void createNotificationAddMix() {
>         Notification n = new Notification();
>         n.setId("add-Mix");
>         NativeButton button1 = new NativeButton(BUTTON_CAPTION);
>         NativeButton button2 = new NativeButton("3333");
>         NativeButton button3 = new NativeButton("4444");
>         NativeButton button4 = new NativeButton("BYE");
>         button1.setId("Add-Mix-open");
>         button2.setId("add-Mix-two");
>         button3.setId("add-Mix-three");
>         button4.setId("add-Mix-close");
>  
>         n.add(button2);
>         n.setText("55555555");
>         n.add(button3);
>         button1.addClickListener(event -> n.open());
>         button4.addClickListener(event -> n.close());
>         add(n, button1, button4);
>     }


You have tree new methods here but the IT test does check the same thing using different identifiers.

I don't see in the tests checks for text presence in the dialog.
</blockquote></details>

added a test for checking text.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 112 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJV2gD-9I-p4CBlbj:-L1WZnsbayFPTpcIYrTo:bb7f2mf) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L112)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> NotificationAddTwoComponents

use camel case method names even for test methods (the first letter).
</blockquote></details>

done

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 122 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJelCPFOz38SlIJmY:-L1WZfrlqWPK4u4vOV1T:bbsr9gr) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L122)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  findElement(By.id("Add-two-components-open")).click();
>         waitUntil(driver -> Boolean.TRUE.toString()
>                 .equals(findElement(By.tagName(DIALOG_OVERLAY_TAG))
>                         .getAttribute("opened")));
>         Assert.assertEquals(2,
>                 getOverlayContent().findElements(By.tagName("button")).size());
>         clickElementWithJs(findElement(By.id("add-two-components-close")));
>         waitUntil(driver -> Boolean.FALSE.toString()
>                 .equals(findElement(By.tagName(DIALOG_OVERLAY_TAG))
>                         .getAttribute("opened")));


All three new methods have the same source code except two String identifiers which can be used as parameters in the extracted method.

Plus the number of button in the dialog. It also may be a parameter.

Don't duplicate the code. Extract the logic into a method and use parameters there.
</blockquote></details>

i have refactor the code. thanks. 

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 126 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJaaxrjMvYB_CRN4e:-L1WZf5urikKqvU3k2G0:bb7f2mf) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L126)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> NotificationAddMix

same here
</blockquote></details>

done

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java, line 140 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WJc7qkuWDe0vWKRdZ:-L1WZbaR4afYznmCsIK_:b-t56g7p) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/479c555969a87d952e5897a55a6208009106d73d/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPageIT.java#L140)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> NotificationwithTextAndAddComponent

same here
</blockquote></details>

sorry.. didn't notice that.. i will remember... done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 153 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1W_zt9t7JNnrYUGbt4:-L1W_zt9t7JNnrYUGbt5:b-g53gpd) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/main/java/com/vaadin/flow/component/notification/Notification.java#L153)):*
> getElement().getNode().runWhenAttached(
>                 ui -> ui.beforeClientResponse(this, () -> templateElement
>                         .setProperty("innerHTML", HtmlUtils.escape(text))));

I don't see any reason todo it in this way.
Please return back.

`beforeClientResponse` and `runWhenAttached` are normally used to execute JS (which shouldn't be executed if the node is not attached).

`setProperty` logic already does everything for you out out of the box.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTest.java, line 68 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Wal0_3hmSwrUJ4Jex:-L1Wal0_3hmSwrUJ4Jey:b-97jymd) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTest.java#L68)):*
>  @Test
>     public void setContent() {
>         notification = new Notification("new");

Test is removed without a replacement.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java, line 165 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WbE-ZQJNpwVmgtD0M:-L1WbE-ZQJNpwVmgtD0N:bh1laat) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java#L165)):*
> notification.setText("55555555");

You should check that this text is not present (not just the button presence).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 153 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1W_zt9t7JNnrYUGbt4:-L1WbcCoHvHafyDRnheT:bs3y1px) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/main/java/com/vaadin/flow/component/notification/Notification.java#L153)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> getElement().getNode().runWhenAttached(
>                 ui -> ui.beforeClientResponse(this, () -> templateElement
>                         .setProperty("innerHTML", HtmlUtils.escape(text))));

I don't see any reason todo it in this way.
Please return back.

`beforeClientResponse` and `runWhenAttached` are normally used to execute JS (which shouldn't be executed if the node is not attached).

`setProperty` logic already does everything for you out out of the box.
</blockquote></details>

This is added for one use case.. 

If i do in this order, `create a notification`, `add a component`, `add text`, the result is that component has been removed, but text will not set to the template. I think the reason for this is from `getElement().getNode().runWhenAttached(ui -> ui.beforeClientResponse(this, () -> attachComponentTemplate(ui)));`(especially the attachComponentTemplate()) which comes from add component method, because we are using the same template for component and text at the moment.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTest.java, line 68 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Wal0_3hmSwrUJ4Jex:-L1WbvKtxSRRcvxRsqUO:b-o4yol2) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTest.java#L68)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  @Test
>     public void setContent() {
>         notification = new Notification("new");

Test is removed without a replacement.
</blockquote></details>

This test is not suitable for a unit test anymore... so i added the check in IT test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Review status: 5 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java, line 165 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1WbE-ZQJNpwVmgtD0M:-L1Wd9Bot77vptc0cEls:b-ye6hnu) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTestPage.java#L165)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> notification.setText("55555555");

You should check that this text is not present (not just the button presence).
</blockquote></details>

added. thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: 5 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 153 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1W_zt9t7JNnrYUGbt4:-L1Weqk1K1aW04UnEr3K:bn6d031) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/main/java/com/vaadin/flow/component/notification/Notification.java#L153)):*
<details><summary><i>Previously, ZheSun88 (Sun Zhe) wrote…</i></summary><blockquote>

This is added for one use case.. 

If i do in this order, `create a notification`, `add a component`, `add text`, the result is that component has been removed, but text will not set to the template. I think the reason for this is from `getElement().getNode().runWhenAttached(ui -> ui.beforeClientResponse(this, () -> attachComponentTemplate(ui)));`(especially the attachComponentTemplate()) which comes from add component method, because we are using the same template for component and text at the moment.
</blockquote></details>

It's because the implementation is bad.

The logic which does something to show added components is always executed in the very end because it's deferred via `beforeClientResponse`.

The property is set always at the time when the method is called. So `attachComponentTemplate` always wins: it's executed after any property change call.

Instead deferring the property set via `beforeClientResponse` you should do:
- when the text is set all deferred tasks should be canceled
- when a components  is added the text should be reset immediately so that it allows deferred tasks execution.

This can be done in many ways but I suggest: 
- check (somehow, may be via introducing an additional flag for it) whether the text is set in the `attachComponentTemplate` method. It should execute its logic _ONLY_ if the text is not set.
- as mentioned any component addition should somehow resets the text, so that `attachComponentTemplate` is allowed to execute.

And this _HAS_ to have a unit test.

---

*[src/test/java/com/vaadin/flow/component/notification/tests/NotificationTest.java, line 68 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1Wal0_3hmSwrUJ4Jex:-L1WgmZhCuFK3SHceEim:bp1273h) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/13b76ec01066f93a4f21e823b888ff4e03f13c99/src/test/java/com/vaadin/flow/component/notification/tests/NotificationTest.java#L68)):*
<details><summary><i>Previously, ZheSun88 (Sun Zhe) wrote…</i></summary><blockquote>

This test is not suitable for a unit test anymore... so i added the check in IT test
</blockquote></details>

Regardless of how it's implemented there has to be a unit test for this.

It's possible to emulate `runWhenAttached` and `beforeClientResponse`.
("Emulate" is a bad word here. It's possible to call something which leads to execution of those methods so that the result may be checked).
`runWhenAttached` is executed immediately if the node is attached ( you will need a `UI` instance and add a child to it. Such child becomes attached).

`beforeClientResponse` is executed when `StateTree::runExecutionsBeforeClientResponse` is called.

So there are no any obstacles to test this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 153 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1W_zt9t7JNnrYUGbt4:-L1XZdsXyfkIbZTbLEwe:b-kkqgrh) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/main/java/com/vaadin/flow/component/notification/Notification.java#L153)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

It's because the implementation is bad.

The logic which does something to show added components is always executed in the very end because it's deferred via `beforeClientResponse`.

The property is set always at the time when the method is called. So `attachComponentTemplate` always wins: it's executed after any property change call.

Instead deferring the property set via `beforeClientResponse` you should do:
- when the text is set all deferred tasks should be canceled
- when a components  is added the text should be reset immediately so that it allows deferred tasks execution.

This can be done in many ways but I suggest: 
- check (somehow, may be via introducing an additional flag for it) whether the text is set in the `attachComponentTemplate` method. It should execute its logic _ONLY_ if the text is not set.
- as mentioned any component addition should somehow resets the text, so that `attachComponentTemplate` is allowed to execute.

And this _HAS_ to have a unit test.
</blockquote></details>

Please create a ticket to improve the implementation.
Unblocked.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->

ZheSun88 commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/main/java/com/vaadin/flow/component/notification/Notification.java, line 153 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8#-L1W_zt9t7JNnrYUGbt4:-L1X_h4OwpIHJrDm6Fur:b-sw77t0) ([raw file](https://github.com/vaadin/vaadin-notification-flow/blob/eaaac2d93222fb554e28316e60d5ebeb1366bc02/src/main/java/com/vaadin/flow/component/notification/Notification.java#L153)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

Please create a ticket to improve the implementation.
Unblocked.
</blockquote></details>

ticket https://github.com/vaadin/vaadin-notification-flow/issues/10

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-notification-flow/8)*
<!-- Sent from Reviewable.io -->
